### ISSUE TITLE
"WARNING:tensorflow: You are casting an input of type complex64 to an incompatible dtype float32.  This will discard the imaginary part and may not be what you intended."

### ISSUE THREAD

**Issue by fafrincs** [NEGU93/cvnn#35]
The issue I have encountered while using the CVNN library is that I have received warning messages indicating that there is a mismatch between the data types of the inputs to certain operations in my code.

"WARNING:tensorflow: You are casting an input of type complex64 to an incompatible dtype float32.  This will discard the imaginary part and may not be what you intended."

 It seems like the issue may be related to using complex-valued inputs with operations that expect real-valued inputs. But this library is intended to work with complex data types that is what the goal is. Any idea how can I fix this problem?
![dd](https://user-images.githubusercontent.com/116988923/235208384-97c58a10-bd12-4773-b385-d301531a4388.JPG)

The attached image is the output of CIFAR 10 examples.  https://complex-valued-neural-networks.readthedocs.io/en/latest/code_examples/cifar10.html 

I would appreciate any kind of help I receive.

**Comment by nipulseervi** [comment]
remove dype  or change it to tf.complex64

eg:
import tensorflow as tf
tf. device("gpu:1")
model =  tf.keras.models.Sequential()
model.add(complex_layers.ComplexInput(input_shape=input_shape))
model.add(complex_layers.ComplexConv2D(32, (2, 2), activation='cart_leaky_relu',  padding='same'))
model.add(complex_layers.ComplexMaxPooling2D((2, 2)))
model.add(complex_layers.ComplexConv2D(64, (2, 2), activation='cart_leaky_relu',  padding='same')) 
model.add(complex_layers.ComplexMaxPooling2D((2, 2)))
model.add(complex_layers.ComplexConv2D(128, (2, 2), activation='cart_leaky_relu',  padding='same'))
model.add(complex_layers.ComplexMaxPooling2D((2, 2)))
model.add(complex_layers.ComplexConv2D(256, (2, 2), activation='cart_leaky_relu',  padding='same'))
model.add(complex_layers.ComplexMaxPooling2D((2, 2)))
model.add(complex_layers.ComplexFlatten())

**Comment by fafrincs** [comment]
@nipulseervi  Thank you for your previous help. I've successfully removed the dtype and modified the activations to "cart_leaky_relu", it removed the warning. However, I'm wondering if this "cvnn" library other activation functions can handle complex numbers directly without considering the real and imaginary parts separately.
I've tried using activation functions like modrelu, crelu, and zrelu, but I keep receiving the following warning:
“WARNING:tensorflow:You are casting an input of type complex64 to an incompatible dtype float32. This will discard the imaginary part and may not be what you intended.”
I would like to know if these activations can handle complex numbers directly, preserving both the real and imaginary parts. I would appreciate any guidance on how to use these models.

**Comment by maorgranot1** [comment]
this solution might not be efficient but this change solved this error for 'modrelu'
within cvnn/activations.py change the function modrelu to:
"
def modrelu(z: Tensor, b: float = 1., c: float = 1e-3) -> Tensor:
    abs_z = tf.math.abs(z)
    mod_relu = tf.keras.activations.relu(abs_z + b)
    return tf.complex(mod_relu * tf.math.real(z) / (abs_z + c), mod_relu * tf.math.imag(z) / (abs_z + c))
"

**Comment by NEGU93** [comment]
Sorry for my late reply.
This error was because of your line that uses `'relu'` activation function. Here you are using Tensorflow ReLU. Tensorflow is not prepared to deal with complex values so it cast it to float. My library then realizes this happens and because I cannot go to Tensorflow to change stuff I at least warn you about this. Please use only activation functions from cvnn toolbox or code an activation function yourself.

**Comment by LiYongNing98** [comment]
> @nipulseervi Thank you for your previous help. I've successfully removed the dtype and modified the activations to "cart_leaky_relu", it removed the warning. However, I'm wondering if this "cvnn" library other activation functions can handle complex numbers directly without considering the real and imaginary parts separately. I've tried using activation functions like modrelu, crelu, and zrelu, but I keep receiving the following warning: “WARNING:tensorflow:You are casting an input of type complex64 to an incompatible dtype float32. This will discard the imaginary part and may not be what you intended.” I would like to know if these activations can handle complex numbers directly, preserving both the real and imaginary parts. I would appreciate any guidance on how to use these models.

Hello！May I ask you a question? It has been a year since you last asked this question. Also I would like to know if 'cvnn' is able to handle plural data directly. I am very much looking forward to your reply! Have a nice life!
